### PR TITLE
Update plugins architecture

### PR DESCRIPTION
**Pull request by olivierodo**
Due to major changes on the plugin architecture (see  #112).
We will reshape how plugin should be load into RestQA.

- [x] Refactor bootstrap.js
- [x] Fix side effect on the `restqa steps` command line
- [x] Increase the integration test

### CODE REVIEW

File: src/core/data.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    if (this._options.startSymbol.length !== 2) {
+      throw new Error("The startSymbol should contains 2 charaters");
+    }
+
+    if (this._options.endSymbol.length !== 2) {
+      throw new Error("The endSymbol should contains 2 charaters");
+    }
+
+    const o = {
```
Comment (tony-go) [inline src/core/data.js:21]: 🤓 I'm not against abbreviations in some cases (as iterator callbacks) but here it's really hard to guess what it?
Moreover, I think we could just use two variables (`start` and `end` ) as `o` is never used alone.

File: src/core/data.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      .map((_) => _.replace(/.[^.]*$/, "")); // return an array like ['1.users']
+
+    list = [...new Set(list)]; // dedup array
+    const result = this.fetch(list);
+    for await (const entity of result) {
+      this.set(entity.key, entity.value);
+    }
+  }
+
+  async *fetch(list) {
```
Comment (tony-go) [inline src/core/data.js:63]: ❤️

File: src/core/data.test.js [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    const data = new Data(options);
+    data.set("foo", "bar");
+    expect(data.get("foo")).toEqual("foo");
+    expect(data.get("{{foo")).toEqual("{{foo");
+    expect(data.get("{{ foo}}")).toEqual("bar");
+    expect(data.get("{{ foo }}")).toEqual("bar");
+    expect(data.get("{{foo }}")).toEqual("bar");
+    expect(data.get("{{foo}}")).toEqual("bar");
+    expect(data.get("{{foo o }}")).toEqual("{{foo o }}");
+  });
```
Comment (tony-go) [inline src/core/data.test.js:87]: We probably have two cases there. 
- when the key exists we get the value
- when the key didn't exist (or match) we got the key back.

Or at least we can use named expected results to clarify:
```suggestion
  test("add/get values to the datastore (string)", () => {
    const options = {};
    const data = new Data(options);
    const key = "foo";
    const value = "bar";
  
    data.set(key, value);
  
    // Return back the key passed in args
    const wrongKey = "foo";
    expect(data.get(wrongKey)).toEqual(wrongKey);
    const wrongKey2 = "{{foo";
    expect(data.get(wrongKey2)).toEqual(wrongKey2);
    // on so on ...
    expect(data.get("{{foo o }}")).toEqual("{{foo o }}");
    
    // Good cases
    expect(data.get("{{ foo}}")).toEqual(value);
    expect(data.get("{{ foo }}")).toEqual(value);
    expect(data.get("{{foo }}")).toEqual(value);
    expect(data.get("{{foo}}")).toEqual(value);
  });
```

File: src/core/data.test.js [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    expect(data.get("{{foo")).toEqual("{{foo");
+    expect(data.get("{{ foo}}")).toEqual(2021);
+    expect(data.get("{{ foo }}")).toEqual(2021);
+    expect(data.get("{{foo }}")).toEqual(2021);
+    expect(data.get("{{foo}}")).toEqual(2021);
+    expect(data.get("{{foo}} year")).toEqual("2021 year");
+    expect(data.get("{{foo o }} Nice")).toEqual("{{foo o }} Nice");
+  });
+
+  test("Do not parse the data from the scenario since the there is no channel associated", () => {
```
Comment (tony-go) [inline src/core/data.test.js:103]: Is no channel related to no options?

Comment (tony-go) [reply src/core/data.test.js:103]: I think it's more something like `{ channel: undefined }` that we need here. It's add a bit of expressiveness.

**Review by tony-go — CHANGES_REQUESTED** [review]
👏 Awesome! 

I left a lot of comments 🙈, I know that you put a lot of effort into this. Feel free to reject my suggestions

File: src/core/bootstrap.test.js [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+      url: "https://api.restqa.io"
+    });
+
+    expect(processor.Given).toHaveBeenCalledTimes(1);
+    expect(processor.Given.mock.calls[0][0]).toEqual("my step");
+
+    expect(processor.defineParameterType).toHaveBeenCalledTimes(1);
+
+    expect(processor.setWorldConstructor).toHaveBeenCalledTimes(1);
+    const World = processor.setWorldConstructor.mock.calls[0][0];
```
[the remark this answers is not in the document]
Comment (olivierodo) [reply src/core/bootstrap.test.js:126]: I'm not sure how to do this one 🥲

File: src/core/bootstrap.test.js [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+    const {regexp, transformer, name} =
+      processor.defineParameterType.mock.calls[0][0];
+    expect(regexp).toEqual(/\{\{(.*)\}\}/);
+    expect(name).toEqual("data");
+    expect(transformer.call(world, "foo")).toEqual("bar");
+
+    expect(jestqa.getLoggerMock()).toHaveBeenCalledTimes(1);
+    expect(jestqa.getLoggerMock().mock.calls[0][0]).toMatch(
+      '🎯 The selected environment is: "local"'
+    );
```
[the remark this answers is not in the document]
Comment (olivierodo) [reply src/core/bootstrap.test.js:143]: Actually on this one we need to use `toMatch` becausre the string is has invisible character in order to colorize the output in the terminal

File: src/core/bootstrap.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (tony-go) [reply src/core/bootstrap.test.js:131]: Me neither aahha

File: src/core/bootstrap.test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (tony-go) [reply src/core/bootstrap.test.js:148]: Maybe could we use this ? https://jestjs.io/docs/expect#tomatchregexp--string
```suggestion
expect(jestqa.getLoggerMock()).toHaveBeenCalledWith(expect.toMatch(
      '🎯 The selected environment is: "local"')
    );
```

I know that it's work with `expect.toMatchObject`

**Review by tony-go — APPROVED** [review]

**Pull request merged**